pyirf/interpolation/moment_morph_interpolator.py

```diff
@@ -0,0 +1,354 @@
+import numpy as np
+from pyirf.binning import bin_center
+from scipy.spatial import Delaunay
+
+from .base_interpolators import DiscretePDFInterpolator
+
+__all__ = [
+    "MomentMorphInterpolator",
+]
+
+
+def estimate_mean_std(bin_edges, bin_contents):
+    """
+    Function to roughly estimate mean and standard deviation from a histogram.
+
+    Parameters
+    ----------
+    bin_edges: np.ndarray, shape=(M+1)
+        Array of common bin-edges for bin_contents
+    bin_contents: np.ndarray, shape=(N, ..., M)
+        Array of bin-entries, actual
+
+    Returns
+    -------
+    mean: np.ndarray, shape=(N, ...)
+        Estimated mean for each input template
+    std: np.ndarray, shape=(N, ...)
+        Estimated standard deviation for each input template. Set to width/2 if only one bin in
+        the input template is =/= 0
+    """
+    # Create an 2darray where the 1darray mids is repeated n_template times
+    mids = np.broadcast_to(bin_center(bin_edges), bin_contents.shape)
+    # Weighted averages to compute mean and std
+    mean = np.average(mids, weights=bin_contents, axis=-1)
+    std = np.sqrt(
+        np.average((mids - mean[..., np.newaxis]) ** 2, weights=bin_contents, axis=-1)
+    )
+
+    # Set std to 0.5*width for all those templates that have only one bin =/= 0. In those
+    # cases mids-mean = 0 and therefore std = 0. Uses the width of the one bin with
+    # bin_content!=0
+    mask = std == 0
+    if np.any(mask):
+        # Create array of bin_widths and clone for each template
+        width = np.diff(bin_edges) / 2
+        width = np.repeat(width[np.newaxis, :], bin_contents.shape[0], axis=0)
+        std[mask] = width[bin_contents[mask, :] != 0]
+
+    return mean, std
+
+
+def lookup(bin_edges, bin_contents, x):
+    """
+    Function to return the bin-height at a desired point.
+
+    Parameters
+    ----------
+    bin_edges: np.ndarray, shape=(M+1)
+        Array of common bin-edges for bin_contents
+    bin_contents: np.ndarray, shape=(N, ..., M)
+        Array of bin-entries, actual
+    x: numpy.ndarray, shape=(N, ..., M)
+        Array of M points for each input template, where the histogram-value (bin-height) should be found
+
+    Returns
+    -------
+    y: numpy.ndarray, shape=(N, ..., M)
+        Array of the bin-heights at the M points x, set to 0 at each point outside the histogram
+
+    """
+    # Create a flattend version of self.bin_contents to ease broadcasting
+    intermediate_bin_contentents = bin_contents.reshape(-1, bin_contents.shape[-1])
+
+    # Find the bin where each point x is located in
+    binnr = np.digitize(x, bin_edges).reshape(-1, bin_contents.shape[-1])
+
+    # np.digitize returns 0 if x is below the first and len(bins) if x is above the last bin,
+    # set these to 0 in a copy of the binnr to avoid errors later
+    binnr_copy = np.copy(binnr)
```

Comment:
why copy?

It's not an input, so ok to mutate. also, where should create a fresh array again.